src/openfermion/utils/_sparse_tools.py

```diff
@@ -250,6 +250,68 @@ def jw_number_restrict_operator(operator, n_electrons, n_qubits=None):
     return operator[numpy.ix_(select_indices, select_indices)]
 
 
+def jw_get_ground_states_by_particle_number(sparse_operator, particle_number):
+    """For a Jordan-Wigner encoded operator, compute the lowest eigenvalue and
+    eigenstates at a particular particle number. The operator must conserve
+    particle number.
+
+    Args:
+        sparse_operator: A Jordan-Wigner encoded sparse operator.
+        particle_number: The particle number at which to compute ground states.
+
+    Returns:
+        ground_energy: The lowest eigenvalue of sparse_operator within the
+            eigenspace of the number operator corresponding to particle_number.
+        ground_states: A list of the corresponding eigenstates.
+
+    Warning: The running time of this method is exponential in the number
+        of qubits.
+    """
+    # Check if operator is Hermitian
+    if not is_hermitian(sparse_operator):
+        raise ValueError('sparse_operator must be Hermitian.')
+
+    n_qubits = int(numpy.log2(sparse_operator.shape[0]))
+
+    # Check if operator conserves particle number
+    sparse_num_op = jordan_wigner_sparse(number_operator(n_qubits))
+    com = commutator(sparse_num_op, sparse_operator)
+    if com.nnz:
+        maxval = max(map(abs, com.data))
+        if maxval > EQ_TOLERANCE:
+            raise ValueError('sparse_operator must conserve particle number.')
+
+    # Get the operator restricted to the sector of the desired particle number,
+    # and compute its eigenvalues and eigenvectors
+    restricted_operator = jw_number_restrict_operator(sparse_operator,
+                                                      particle_number,
+                                                      n_qubits)
+    dense_restricted_operator = restricted_operator.toarray()
+    eigvals, eigvecs = numpy.linalg.eigh(dense_restricted_operator)
```

Comment:
I might be missing something here but it looks like you are doing a dense diagonalization and then returning all of the eigenvectors and values associated with the nonzero eigenvalues. In this case, the function is not really about ground states, it is about all the eigenstates, right?

If you did want to specialize it to the low energy part of the spectrum I would think you would want to keep things sparse and use scipy.sparse.linalg which uses the lanczos routine (much faster when just trying to get the extremal eigenvalues of a sparse matrix).

Reply:
I was thinking about the case that the ground space within a particle number sector was degenerate and spanned the whole sector, then we wouldn't get all the eigenvectors. I hadn't thought much about a better way to get around that or if it's an issue.

Reply:
Okay so it sounds like your intention is function that just gets the low energy nontrivial eigenvalues (so you are after the ground state) but you are worried about the case when there is only one nonzero eigenvalue and many associated eigenvectors? I do not think that can happen.

But I realize that there is a different problem with using lanczos in this situation. The problem is that if the lowest nontrivial eigenvalue is greater than zero you are in trouble. Because then the lowest ones will all be zeros.

I would perhaps suggest keeping your code as it is but reselling it as code to get the nontrivial part of the entire spectrum (so all eigenstates and eigenvalues that are not zero eigenvalue). Perhaps call the function "number_restricted_diagonalization" or something like that. Would that be okay?

Reply:
Sorry, I don't understand the problem you pointed out with lanczos. It seems to me that if we used it instead of the dense routine, we would still get the same eigenvalue that I denoted `ground_energy`. The ground energy generally refers to the lowest eigenvalue, even if it's 0, right?

Reply:
Ah wait, I should have read your first comment more clearly. This function returns the lowest eigenvalue, not the nonzero eigenvalues.

Reply:
At least that's what it's supposed to do!

Reply:
Oh right, sorry (a little tired over here). I was thinking for some reason that the jw_number_restrict code just projects out parts of the operator in the wrong particle sector. Were that the case, you'd be left with an operator that has a bunch of trivial zero eigenvalues. So that is what I was complaining about. But I realize now that it actually gives you the smaller operator.

So indeed, you should be fine using lanczos. Does that work for you? I would not be too worried about a bunch of degenerate eigenvalues. Perhaps you can request that lanczos return the lowest three or four eigenvalues and raise a warning if by some chance they are all the same (the warning would be that there might be yet more generate eigenvalues). Perhaps you can have an optional parameter which switches between the dense and sparse diagonalization.

How does that sound?

Reply:
I think that's a great idea!